mava/systems/q_learning/rec_iql.py

```diff
@@ -0,0 +1,662 @@
+
+        # Scan over each sample and discard first timestep
+        hidden_state, obs_done = prep_inputs_to_scannedrnn(data.obs, data.done)
+
+        _, next_online_greedy_dist = q_net.apply(
+            params.online, hidden_state, obs_done
+        )  # eps defaults to 0
+
+        _, next_q_vals_target = q_net.apply(
+            params.target, hidden_state, obs_done, method="get_q_values"
+        )
+
+        # Skip first timestep, since we are working with "next" information
+        next_action = next_online_greedy_dist.mode()[1:, ...]  # (T, B, ...)
+        next_q_vals_target = next_q_vals_target[1:, ...]  # (T, B, ...)
+
+        # Double q-value selection
+        next_q_val = jnp.squeeze(
+            jnp.take_along_axis(next_q_vals_target, next_action[..., jnp.newaxis], axis=-1), axis=-1
+        )
+
+        next_q_val = switch_leading_axes(next_q_val)  # (T, B, ...) -> (B, T, ...)
+
+        # TD Target
+        target_q_val = (
+            first_reward
+            + (1.0 - jnp.array(next_done, dtype="float32")) * cfg.system.gamma * next_q_val
+        )
+
+        # Update Q function.
+        q_grad_fn = jax.grad(q_loss_fn, has_aux=True)
+        q_grads, q_loss_info = q_grad_fn(
+            params.online, data_first.obs, data_first.done, data_first.action, target_q_val
+        )
+
+        # Mean over the device and batch dimension.
+        q_grads, q_loss_info = lax.pmean((q_grads, q_loss_info), axis_name="device")
+        q_grads, q_loss_info = lax.pmean((q_grads, q_loss_info), axis_name="batch")
+        q_updates, next_opt_state = opt.update(q_grads, opt_states)
+        next_online_params = optax.apply_updates(params.online, q_updates)
+
+        target_update_params = (next_online_params, params.target, t_train)
+
+        # Depending on choice of target network strategy, update target weights.
+        next_target_params = jax.lax.cond(
+            cfg.system.hard_update,
+            hard_update,
+            soft_update,
+            *target_update_params,
+        )
+
+        # Repack params and opt_states.
+        next_params = DDQNParams(next_online_params, next_target_params)
+
+        return next_params, next_opt_state, q_loss_info
+
+    # TRAIN LEVEL 1
+    def train(train_state: TrainState, _: Any) -> Tuple[TrainState, Metrics]:
+        """Sample, train and repack."""
+
+        # unpack and get keys
+        buffer_state, params, opt_states, t_train, key = train_state
+        next_key, buff_key = jax.random.split(key, 2)
+
+        # sample
+        data = rb.sample(buffer_state, buff_key).experience
+
+        # learn
+        next_params, next_opt_states, q_loss_info = update_q(params, opt_states, data, t_train)
+
+        # Repack.
+        next_train_state = TrainState(
+            buffer_state, next_params, next_opt_states, t_train + 1, next_key
+        )
+
+        return next_train_state, q_loss_info
+
+    # INTERACT-TRAIN LOOP
+
+    scanned_interact = lambda state: lax.scan(
```

Comment:
I changed act to interact in my code because I found act to not be representative of what the function is doing. My interact function selects, steps, then stores. I am willing to change it, but I will say it's not my preference!